### PR TITLE
Change `Rc::inc_{weak,strong}` to better hint optimization to LLVM

### PR DESCRIPTION
**Pull request by dshynkev**
As discussed in #13018, `Rc::inc_strong` and `Rc::inc_weak` are changed to allow compositions of `clone` and `drop` to be better optimized. Almost entirely as in [this comment](https://github.com/rust-lang/rust/issues/13018#issuecomment-408642184), except that `abort` on zero is added so that a `drop(t.clone())` does not produce a zero check followed by conditional deallocation.

This is different from #21418 in that it doesn't rely on `assume`, avoiding the prohibitive compilation slowdown.

[Before and after IR](https://gist.github.com/hermord/266e55451b7fe0bb8caa6e35d17c86e1).

### CODE REVIEW

File: src/liballoc/rc.rs [hunk 1]
```diff
@@ -1359,7 +1359,14 @@ trait RcBoxPtr<T: ?Sized> {
 
     #[inline]
     fn inc_strong(&self) {
-        self.inner().strong.set(self.strong().checked_add(1).unwrap_or_else(|| unsafe { abort() }));
+        // We want to abort on overflow instead of dropping the value.
+        // The reference count will never be zero when this is called;
+        // nevertheless, we insert an abort here to hint LLVM at
+        // an otherwise missied optimization.
+        if self.strong() == 0 || self.strong() == usize::max_value() {
```
Comment (Mark-Simulacrum) [inline src/liballoc/rc.rs]: Hm, should we instead have unreachable_unchecked for the == 0 case?

File: src/liballoc/rc.rs [hunk 2]
```diff
@@ -1359,7 +1359,14 @@ trait RcBoxPtr<T: ?Sized> {
 
     #[inline]
     fn inc_strong(&self) {
-        self.inner().strong.set(self.strong().checked_add(1).unwrap_or_else(|| unsafe { abort() }));
+        // We want to abort on overflow instead of dropping the value.
+        // The reference count will never be zero when this is called;
+        // nevertheless, we insert an abort here to hint LLVM at
+        // an otherwise missied optimization.
```
Comment (theotherphil) [inline src/liballoc/rc.rs]: Typo: “missied” (here and below).

File: src/liballoc/rc.rs (hunk 1 above)
Comment (hermord) [reply src/liballoc/rc.rs]: Confusingly, doing this:
```diff
-        if self.strong() == 0 || self.strong() == usize::max_value() {
+        if self.strong() == 0 {
+            unsafe { unreachable_unchecked(); }
+        }
+        if self.strong() == usize::max_value() {
             unsafe { abort(); }
```
produces the same output as before any changes (`before` in the linked comparison).

**Comment by TimNN** [comment]
Ping from triage @aidanhs. This PR requires your review.

**Comment by hermord** [comment]
@alexcrichton The code currently doesn't check for it, but it's needed to hint LLVM that the `clone -> drop` seqeunce will not bring the reference count to zero. #21418 did this with `assume`, but that turned out to be prohibitively expensive. I tried hinting this with `unreachable_unchecked`, as @Mark-Simulacrum suggested in an inline comment upthread, but, for some reason, that hint is discarded. Your comment actually prompted me to investigate this on the LLVM side, but I've been unsuccessful so far.

**Comment by alexcrichton** [comment]
Ok looks and sounds great to me, thanks!

@bors: r+

**Comment by kennytm** [comment]
@bors r-

You have accidentally changed the `src/tools/lld` submodule.

**Comment by eddyb** [comment]
You might be able to undo the problem by doing `git checkout HEAD~1 src/tools/lld`.

**Comment by hermord** [comment]
Oops, my apologies. That's what I get for doing `git stage -u`.

**Comment by hermord** [comment]
Fixed now.

**Pull request merged**